File: global-hooks/framework/memory/auto_memory_writer.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
auto_memory_writer.py - Stop Hook (Layer 2: Project Mid-Term Memory Write)
===========================================================================

At session end, appends a compact summary to .claude/MEMORY.md.

Captures what actually happened in the session using ground-truth signals:
  - git diff --stat (files changed — most reliable signal)
  - git log -1 (last commit message if one was made)
  - Task summaries from auto_context_manager compressed context

Format: one dated entry per session, ~5-15 lines.
Only writes if something actually changed (git diff non-empty).

Exit: always 0 (never blocks)
"""

import json
import os
import re
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

COMPRESSED_DIR = Path.home() / ".claude" / "data" / "compressed_context"
MAX_MEMORY_ENTRIES = 30     # Prune oldest entries if file grows beyond this
MAX_ENTRY_LINES = 20        # Cap per-session entry to keep file lean


def memory_path(cwd: str) -> Path:
    return Path(cwd).resolve() / ".claude" / "MEMORY.md"


def run(cmd: list, cwd: str, timeout: int = 5) -> str:
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, cwd=cwd, timeout=timeout)
        return r.stdout.strip() if r.returncode == 0 else ""
    except Exception:
        return ""


def get_git_changes(cwd: str) -> list[str]:
    """Get changed files from git diff. Returns list of 'N files changed' lines."""
    stat = run(["git", "diff", "--stat", "HEAD~1", "HEAD", "--"], cwd)
    if not stat:
        # No commit yet — try staged/unstaged changes
        stat = run(["git", "diff", "--stat"], cwd)
    if not stat:
        stat = run(["git", "diff", "--stat", "--cached"], cwd)
    if not stat:
        return []
    return [line for line in stat.splitlines() if line.strip()]


def get_last_commit(cwd: str) -> str:
    return run(["git", "log", "-1", "--format=%s (%h)"], cwd)


def get_compressed_summaries(session_id: str) -> list[str]:
    """Read any compressed task summaries from auto_context_manager for this session."""
    summaries = []
    if not COMPRESSED_DIR.exists():
        return summaries
    try:
        for f in sorted(COMPRESSED_DIR.glob(f"{session_id[:8]}*.json")):
            data = json.loads(f.read_text())
            subj = data.get("subject", "")
            outcome = data.get("outcome", "")
            if subj:
                line = f"- Task: {subj}"
                if outcome:
                    line += f" → {outcome}"
                summaries.append(line)
    except Exception:
        pass
    return summaries[:5]  # Cap at 5 task summaries


def build_entry(cwd: str, session_id: str) -> str | None:
    """Build a session memory entry. Returns None if nothing to record."""
    lines = []
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    time_str = datetime.now(timezone.utc).strftime("%H:%M UTC")
    project = Path(cwd).resolve().name

    # Ground truth: what files actually changed
    git_changes = get_git_changes(cwd)
    last_commit = get_last_commit(cwd)
    task_summaries = get_compressed_summaries(session_id)

    # Only write if something happened
    if not git_changes and not task_summaries:
        return None

    lines.append(f"## {today} ({time_str})")

    if last_commit:
        lines.append(f"**Commit:** {last_commit}")

    if git_changes:
        lines.append("**Changed:**")
        # Show file list, cap at 10
        for line in git_changes[:10]:
            if "|" in line or "changed" in line:
                lines.append(f"  {line.strip()}")
        if len(git_changes) > 10:
            lines.append(f"  ... and {len(git_changes) - 10} more files")

    if task_summaries:
        lines.append("**Tasks completed:**")
        lines.extend(task_summaries)

    lines.append("")  # trailing newline
    return "\n".join(lines)


def prune_old_entries(content: str, max_entries: int) -> str:
    """Keep only the most recent max_entries session blocks."""
    parts = re.split(r"(## \d{4}-\d{2}-\d{2})", content)
    if len(parts) <= 1:
        return content

    file_header = parts[0]
    sessions = []
    i = 1
    while i < len(parts):
        header = parts[i]
        body = parts[i + 1] if i + 1 < len(parts) else ""
        sessions.append(header + body)
        i += 2

    if len(sessions) <= max_entries:
        return content

    # Keep most recent
    return file_header + "".join(sessions[-max_entries:])


def ensure_memory_file(path: Path, project: str) -> str:
    """Create MEMORY.md if it doesn't exist."""
    if path.exists():
        return path.read_text()

    path.parent.mkdir(parents=True, exist_ok=True)
    header = (
        f"# Project Memory — {project}\n"
        "<!-- Mid-term project memory: one entry per session. Auto-maintained. -->\n"
        "<!-- Layer 2 (episodic): what changed, was fixed, was decided across sessions. -->\n\n"
    )
    path.write_text(header)
    return header


def main():
    try:
        data = json.loads(sys.stdin.read())
        cwd = data.get("cwd", os.getcwd())
        session_id = data.get("session_id", "unknown")
        project = Path(cwd).resolve().name
        path = memory_path(cwd)

        entry = build_entry(cwd, session_id)
        if not entry:
            print(json.dumps({}))
            sys.exit(0)

        content = ensure_memory_file(path, project)
        content = content.rstrip() + "\n\n" + entry
        content = prune_old_entries(content, MAX_MEMORY_ENTRIES)
        path.write_text(content)

    except Exception as e:
        print(f"auto_memory_writer error: {e}", file=sys.stderr)

    print(json.dumps({}))
    sys.exit(0)


if __name__ == "__main__":
    main()

File: caf-hooks/src/hooks/auto_memory_writer.rs
/// Auto Memory Writer — Stop hook.
///
/// Python equivalent: global-hooks/framework/memory/auto_memory_writer.py
///
/// Behavior:
/// - Reads session_id and cwd from stdin JSON
/// - Runs `git diff --stat HEAD~1 HEAD --` (falls back to staged/unstaged diffs)
/// - Runs `git log -1 --format=%s (%h) by %an` for last commit
/// - Reads compressed_context JSON files for task summaries
/// - Builds a dated session entry (skips if nothing changed)
/// - Upserts entry into .claude/MEMORY.md (dedup on commit hash)
/// - Prunes oldest entries if >30 entries
/// - Atomically writes MEMORY.md
/// - Always exits 0 (never blocks)
use std::path::{Path, PathBuf};
use std::process::Command;

use chrono::Utc;
use serde_json::Value;

use crate::io::{atomic_write, read_stdin_value, write_output};
use crate::state::compressed_context_dir;
use crate::types::HookOutput;

const MAX_MEMORY_ENTRIES: usize = 30;

// ---------------------------------------------------------------------------
// Path helpers
// ---------------------------------------------------------------------------

fn memory_path(cwd: &str) -> PathBuf {
    PathBuf::from(cwd).join(".claude").join("MEMORY.md")
}

// ---------------------------------------------------------------------------
// Git subprocess helpers
// ---------------------------------------------------------------------------

fn run_git(args: &[&str], cwd: &str) -> String {
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output();
    match output {
        Ok(o) if o.status.success() => {
            String::from_utf8_lossy(&o.stdout).trim().to_string()
        }
        _ => String::new(),
    }
}

fn get_git_changes(cwd: &str) -> Vec<String> {
    // Try committed changes first
    let stat = run_git(&["diff", "--stat", "HEAD~1", "HEAD", "--"], cwd);
    let stat = if stat.is_empty() {
        // No commit yet — try staged/unstaged changes
        let s = run_git(&["diff", "--stat"], cwd);
        if s.is_empty() {
            run_git(&["diff", "--stat", "--cached"], cwd)
        } else {
            s
        }
    } else {
        stat
    };

    if stat.is_empty() {
        return Vec::new();
    }
    stat.lines()
        .filter(|l| !l.trim().is_empty())
        .map(|l| l.to_string())
        .collect()
}

fn get_commit_hash(cwd: &str) -> String {
    run_git(&["log", "-1", "--format=%h"], cwd)
}

fn get_last_commit(cwd: &str) -> String {
    run_git(&["log", "-1", "--format=%s (%h) by %an"], cwd)
}

fn get_git_author(cwd: &str) -> String {
    let name = run_git(&["config", "user.name"], cwd);
    if !name.is_empty() {
        return name;
    }
    let email = run_git(&["config", "user.email"], cwd);
    if email.is_empty() {
        return String::new();
    }
    email.split('@').next().unwrap_or("").to_string()
}

// ---------------------------------------------------------------------------
// Compressed context summaries
// ---------------------------------------------------------------------------

fn get_compressed_summaries(session_id: &str) -> Vec<String> {
    let mut summaries = Vec::new();
    let dir = compressed_context_dir();
    if !dir.exists() {
        return summaries;
    }

    let prefix = &session_id[..session_id.len().min(8)];
    let pattern = format!("{}*.json", prefix);

    let glob_pattern = dir.join(&pattern);
    let glob_str = glob_pattern.to_string_lossy();

    let mut paths: Vec<PathBuf> = glob::glob(&glob_str)
        .map(|paths| paths.flatten().collect())
        .unwrap_or_default();
    paths.sort();

    for path in paths {
        if let Ok(text) = std::fs::read_to_string(&path) {
            if let Ok(data) = serde_json::from_str::<Value>(&text) {
                let subj = data.get("subject").and_then(|v| v.as_str()).unwrap_or("");
                let outcome = data.get("outcome").and_then(|v| v.as_str()).unwrap_or("");
                if !subj.is_empty() {
                    let line = if outcome.is_empty() {
                        format!("- Task: {}", subj)
                    } else {
                        format!("- Task: {} → {}", subj, outcome)
                    };
                    summaries.push(line);
                }
            }
        }
        if summaries.len() >= 5 {
            break;
        }
    }
    summaries
}

// ---------------------------------------------------------------------------
// Entry builder
// ---------------------------------------------------------------------------

fn build_entry(cwd: &str, session_id: &str) -> Option<String> {
    let today = Utc::now().format("%Y-%m-%d").to_string();
    let time_str = Utc::now().format("%H:%M UTC").to_string();
    let project = Path::new(cwd)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown")
        .to_string();

    let git_changes = get_git_changes(cwd);
    let last_commit = get_last_commit(cwd);
    let task_summaries = get_compressed_summaries(session_id);
    let author = get_git_author(cwd);

    // Only write if something happened
    if git_changes.is_empty() && task_summaries.is_empty() {
        return None;
    }

    let _ = project; // used in header below for context

    let mut lines: Vec<String> = Vec::new();

    let mut header = format!("## {} ({})", today, time_str);
    if !author.is_empty() {
        header.push_str(&format!(" · @{}", author));
    }
    lines.push(header);

    if !last_commit.is_empty() {
        lines.push(format!("**Commit:** {}", last_commit));
    }

    if !git_changes.is_empty() {
        lines.push("**Changed:**".to_string());
        let display_count = git_changes.len().min(10);
        for line in &git_changes[..display_count] {
            if line.contains('|') || line.contains("changed") {
                lines.push(format!("  {}", line.trim()));
            }
        }
        if git_changes.len() > 10 {
            lines.push(format!("  ... and {} more files", git_changes.len() - 10));
        }
    }

    if !task_summaries.is_empty() {
        lines.push("**Tasks completed:**".to_string());
        lines.extend(task_summaries);
    }

    lines.push(String::new()); // trailing newline

    Some(lines.join("\n"))
}

// ---------------------------------------------------------------------------
// Prune old entries (keep most recent MAX_MEMORY_ENTRIES)
// ---------------------------------------------------------------------------

fn prune_old_entries(content: &str, max_entries: usize) -> String {
    // Split on "## YYYY-MM-DD" date headers
    let re = regex::Regex::new(r"(## \d{4}-\d{2}-\d{2})").unwrap();
    let parts: Vec<&str> = re.split(content).collect();
    let headers: Vec<&str> = re.find_iter(content).map(|m| m.as_str()).collect();

    if headers.is_empty() {
        return content.to_string();
    }

    let file_header = parts[0];
    let mut sessions: Vec<String> = Vec::new();
    for (i, header) in headers.iter().enumerate() {
        let body = if i + 1 < parts.len() {
            parts[i + 1]
        } else {
            ""
        };
        sessions.push(format!("{}{}", header, body));
    }

    if sessions.len() <= max_entries {
        return content.to_string();
    }

    // Keep most recent
    let kept = &sessions[sessions.len() - max_entries..];
    format!("{}{}", file_header, kept.join(""))
}

// ---------------------------------------------------------------------------
// Upsert entry (dedup on commit hash)
// ---------------------------------------------------------------------------

fn upsert_entry(content: &str, entry: &str, commit_hash: &str) -> String {
    if commit_hash.is_empty() {
        // No commit hash to dedup on — just append
        return format!("{}\n\n{}", content.trim_end(), entry);
    }

    let re = regex::Regex::new(r"(## \d{4}-\d{2}-\d{2})").unwrap();
    let parts: Vec<&str> = re.split(content).collect();
    let headers: Vec<&str> = re.find_iter(content).map(|m| m.as_str()).collect();

    if headers.is_empty() {
        return format!("{}\n\n{}", content.trim_end(), entry);
    }

    let file_header = parts[0];
    let mut sessions: Vec<String> = Vec::new();
    for (i, header) in headers.iter().enumerate() {
        let body = if i + 1 < parts.len() {
            parts[i + 1]
        } else {
            ""
        };
        sessions.push(format!("{}{}", header, body));
    }

    // Find existing entry with this commit hash and replace it
    let mut replaced = false;
    for block in sessions.iter_mut() {
        if block.contains(commit_hash) {
            *block = format!("{}\n", entry);
            replaced = true;
            break;
        }
    }

    if replaced {
        let body = sessions.join("");
        format!("{}\n\n{}", file_header.trim_end(), body.trim_end())
            + "\n"
    } else {
        format!("{}\n\n{}", content.trim_end(), entry)
    }
}

// ---------------------------------------------------------------------------
// Ensure MEMORY.md exists
// ---------------------------------------------------------------------------

fn ensure_memory_file(path: &Path, project: &str) -> String {
    if path.exists() {
        return std::fs::read_to_string(path).unwrap_or_default();
    }
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let header = format!(
        "# Project Memory — {}\n\
         <!-- Mid-term project memory: one entry per session. Auto-maintained. -->\n\
         <!-- Layer 2 (episodic): what changed, was fixed, was decided across sessions. -->\n\n",
        project
    );
    let _ = std::fs::write(path, &header);
    header
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

pub fn run() {
    let data: Value = read_stdin_value();

    let cwd = data
        .get("cwd")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| {
            std::env::current_dir()
                .map(|p| p.display().to_string())
                .unwrap_or_else(|_| ".".to_string())
        });

    let session_id = data
        .get("session_id")
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();

    let project = Path::new(&cwd)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("unknown")
        .to_string();

    let path = memory_path(&cwd);

    let entry = match build_entry(&cwd, &session_id) {
        Some(e) => e,
        None => {
            // Nothing happened — skip write
            write_output(&HookOutput::empty());
            return;
        }
    };

    let content = ensure_memory_file(&path, &project);
    let commit_hash = get_commit_hash(&cwd);
    let content = upsert_entry(&content, &entry, &commit_hash);
    let content = prune_old_entries(&content, MAX_MEMORY_ENTRIES);

    let _ = atomic_write(&path, &content);

    write_output(&HookOutput::empty());
}
